src/box/raft.c

```diff
@@ -263,25 +292,23 @@ static void
 box_raft_notify_have_quorum(void)
 {
 	struct raft *raft = box_raft();
+	bool has_healthy_quorum = replicaset_has_healthy_quorum();
+	if (box_raft_election_fencing_paused && has_healthy_quorum)
+		box_raft_election_fencing_resume();
+
 	switch (box_election_mode) {
 	case ELECTION_MODE_MANUAL:
 		/* Quorum loss shouldn't interfere with manual elections. */
-		/*
-		 * TODO: make a manually elected leader step off immediately due
-		 * to fencing.
-		 */
 		assert(!raft->is_cfg_candidate);
+		if (!has_healthy_quorum)
+			box_raft_fence();
 		break;
 	case ELECTION_MODE_CANDIDATE:
-		if (replicaset_has_healthy_quorum()) {
+		if (has_healthy_quorum) {
 			raft_cfg_is_candidate(raft, true);
 		} else if (raft->state == RAFT_STATE_CANDIDATE ||
 			   raft->state == RAFT_STATE_LEADER) {
-			/*
-			 * TODO: make the leader step off immediately due to
-			 * fencing.
-			 */
-			raft_cfg_is_candidate_later(raft, false);
+			box_raft_fence();
```

Comment:
I was wrong. raft_cfg_is_candidate_later is needed here as well as box_raft_fence. Sorry.

Reply:
I didn't notice that box_raft_fence only works on leader. Not on a candidate. And pre-vote has to downgrade candidates as well.

Reply:
Fixed, missed that one too.